README.md
# idleant2

changelog.txt
9-10-2018
- 批量自动购买页面
- 移除幼虫团队加成
- 修复世界加成专精文本不显示正确的数量
- 修复图表文字不更新

7-10-2018
- 很多拼写修复

28-09-2018
- 增加 自动购买者最大倍数
- 增加 可以购买时间和经验

26-09-2018
- 格式化时间严格的新选项
- 修复移动端界面Bug

22-09-2018
- 修复递归购买最大数量
- 修复旅行后自动购买者未激活

21-09-2018
- 新增专精重置
- 添加快捷键 m 和 h 去扭曲分钟和小时

19-09-2018
- 修复跳过按钮 (颜色，不显示)
- 修复单词拼写错误

17-09-2018
- + 50％的世界等级专精变为：减少世界旅行需求（例如：在等级 1上大约 -10％，在等级600上大约 -15％）
- 新增禁用扭曲通知的选项
- 从世界旅行要求中删除了材料（仅限新世界）
- 减少黄蜂和蜜蜂的巢穴的要求
- 抛弃供应世界

16-09-2018
- 虫群声望
- 木头 -> 土壤
- 修复手机界面错误
- 添加了禁用资源结束通知的选项

15-09-2018
- 研究自动购买者，将会忽略扭曲和极端
- 减少研究自动购买者的初始超时，从900s到600s

14-09-2018
- 禁用专精物理学
- PlayFab
- 极端现在需要世界等级 > 15 并且当前世界等级 > 50% 最大等级代替 90%

13-09-2018
- 增加漏掉的名字和简介
- 团队和双胞胎批量升级当每个团队的费用都可以用的时候出现
- 降低了旅行所需的材料
- 极端的研究
- 修复主要工人双倍土壤产出
- 修复世界扭曲自动购买者
- 建筑和工程师的研究更便宜了

12-09-2018
- 新增无限循环检测
- 修复无限循环
- 修复批量购买价格
- 修复敌人没有更多的加成
- 修复: 更换世界后时间没有重置
- 修改左侧分类项的界面

11-09-2018
- 修复世界不保存经验
- 修复学生追随者/采矿混淆
- 修复双胞胎四舍五入
- 修复组织图表没有更新
- 灯泡忽略可重复的研究，如果数量大于最大可买

10-09-2018
- 修复加载错误
- 修复一些拼写错误

09-09-2018
- 最初版本


src/app/model/world.ts
import sample from "lodash-es/sample";
import uniq from "lodash-es/uniq";
import { BaseUnit } from "./baseUnit";
import { BugTypes } from "./bugsTypes";
import { CONSTS } from "./CONSTATS";
import { FullUnit } from "./full-unit";
import { Game } from "./game";
import { Malus } from "./malus";
import { MasteryTypes } from "./masteries/mastery";
import { Price } from "./price";
import { Research } from "./research";
import { STRINGS } from "./strings";
import { Utility } from "./utility";

export class World {
  static prefix = new Array<World>();
  static biome = new Array<World>();
  static suffix = new Array<World>();

  name = "";
  level = new Decimal(1);
  prestige = new Decimal(10);

  //  Productions Bonus/Malus
  productionsBonus = new Array<[BaseUnit, Decimal]>();
  productionsEfficienty = new Array<[BaseUnit, Decimal]>();
  productionsAll = new Array<[BaseUnit, Decimal]>();

  // Unlocked stuff
  startingUnlocked = new Array<Research>();

  // Starting stuff
  //  Unit, Quantity
  startingUnit = new Array<[FullUnit, Decimal]>();

  //  Wining condition
  winContidions = new Array<Price>();
  //  This resources must be zero
  notWinConditions = new Array<Malus>();

  //  Additional bug
  additionalBugs = new Array<BugTypes>();

  constructor(public id = "") {
    if (id !== "") this.name = STRINGS.worlds[this.id];
  }

  setLevel(level: Decimal, game: Game) {
    this.winContidions.push(
      new Price(game.ants.nest, CONSTS.BASE_WIN_CONDITION_OTHER)
    );
    this.additionalBugs.push(BugTypes.ANT);

    this.level = new Decimal(level).floor();

    const multi = this.level.div(5).plus(1);
    this.productionsBonus.forEach(b => (b[1] = b[1].times(multi)));
    this.productionsEfficienty.forEach(b => (b[1] = b[1].times(multi)));
    this.productionsAll.forEach(b => (b[1] = b[1].times(multi)));
    this.startingUnit.forEach(b => (b[1] = b[1].times(multi)));

    this.winContidions.forEach(w => {
      w.price = w.price.times(multi);

      w.price = w.base.winNonLiner
        ? w.price.pow(0.7)
        : w.price.times(Decimal.pow(5, multi));

      w.price = w.price.floor();
    });

    this.prestige = this.level
      .times(10)
      .times(this.level.plus(10).log10())
      .floor();

    if (this.level.gte(5) && this.notWinConditions.length < 1) {
      const malus = [
        game.worldMalus.foodMalus1,
        game.worldMalus.woodMalus1,
        game.worldMalus.crystalMalus1,
        game.worldMalus.scienceMalus1
      ];
      this.notWinConditions.push(sample(malus));
    }

    //  Apply Prestige bonus
    let multiBonus = game.allPrestige.worldPrestige.betterWorlds.quantity.times(
      0.1
    );
    multiBonus = multiBonus.times(
      1 + game.allMateries.getSum(MasteryTypes.BETTER_WORLD) * 0.2
    );
    multiBonus = multiBonus.plus(1);

    if (multiBonus.gt(1)) {
      [
        this.productionsBonus,
        this.productionsEfficienty,
        this.productionsAll
      ].forEach(prod => {
        prod.forEach(p => {
          p[1] = p[1].times(multiBonus);
        });
      });
    }
  }
  canTravel(): boolean {
    this.winContidions.forEach(p => {
      p.canBuy = p.base.quantity.gte(p.price);
    });
    return !(
      this.winContidions.findIndex(p => !p.canBuy) > -1 ||
      this.notWinConditions.findIndex(n => !n.isKilled) > -1
    );
  }
  setMalus() {
    this.notWinConditions.forEach(n => {
      n.quantity = new Decimal(this.level.times(5));
      n.producedBy.find(u => u.rateo.lt(0)).producer.unlock();
      const n2 = n.producedBy.find(u => u.rateo.gt(0)).producer;
      const n3 = n2.producedBy.find(u => u.rateo.gt(0)).producer;

      n2.quantity = n.quantity.div(3.5);

      if (n2.quantity.gt(0.1)) {
        n2.unlock();
        // tslint:disable-next-line:prettier
        n3.quantity = n.quantity.div(1.5).pow(0.9);
        if (n3.quantity.gt(0.1)) {
          n3.unlock();
        } else {
          n3.quantity = new Decimal(0);
        }
      } else {
        n2.quantity = new Decimal(0);
      }
    });
  }
  setGame() {
    this.setMalus();
    this.startingUnit.forEach(s => {
      s[0].unlock();
      s[0].quantity = s[1];
    });
  }

  //#region Save and Load
  getSave(): any {
    return {
      n: this.name,
      l: this.level,
      pb: this.productionsBonus.map(b => [b[0].id, b[1]]),
      pe: this.productionsEfficienty.map(b => [b[0].id, b[1]]),
      pa: this.productionsAll.map(b => [b[0].id, b[1]]),
      wc: this.winContidions.map(w => [w.base.id, w.price]),
      nwc: this.notWinConditions.map(n => n.id),
      adb: this.additionalBugs
    };
  }
  restore(data: any, game: Game) {
    if ("n" in data) this.name = data.n;
    if ("l" in data) this.level = new Decimal(data.l);
    if ("pb" in data) {
      this.productionsBonus = data.pb.map(b => [
        this.findBonus(b[0], game),
        new Decimal(b[1])
      ]);
    }
    if ("pe" in data) {
      this.productionsEfficienty = data.pe.map(b => [
        this.findBonus(b[0], game),
        new Decimal(b[1])
      ]);
    }
    if ("pa" in data) {
      this.productionsAll = data.pa.map(b => [
        this.findBonus(b[0], game),
        new Decimal(b[1])
      ]);
    }
    if ("wc" in data) {
      this.winContidions = data.wc.map(
        w =>
          new Price(game.units.find(u => u.id === w[0]), new Decimal(w[1]), 1)
      );
    }
    if ("nwc" in data) {
      this.notWinConditions = data.nwc.map(nwc =>
        game.units.find(u => u.id === nwc)
      );
    }
    if ("adb" in data) {
      this.additionalBugs = data.adb;
    }
  }
  findBonus(id: string, game: Game): BaseUnit {
    return game.worldBonus.bonusList.find(b => b.id === id);
  }
  //#endregion

  // tslint:disable-next-line:member-ordering
  static merge(worlds: World[]): World {
    const retWorld = new World();

    retWorld.name = worlds.map(w => w.name).reduce((p, c) => p + " " + c);

    worlds.forEach(w => {
      w.productionsBonus.forEach(a => {
        const prod = retWorld.productionsBonus.find(p => p[0].id === a[0].id);
        if (!prod) retWorld.productionsBonus.push([a[0], new Decimal(a[1])]);
        else prod[1] = prod[1].plus(a[1]);
      });
      w.productionsEfficienty.forEach(a => {
        const prod = retWorld.productionsEfficienty.find(
          p => p[0].id === a[0].id
        );
        if (!prod) {
          retWorld.productionsEfficienty.push([a[0], new Decimal(a[1])]);
        } else prod[1] = prod[1].plus(a[1]);
      });
      w.productionsAll.forEach(a => {
        const prod = retWorld.productionsAll.find(p => p[0].id === a[0].id);
        if (!prod) retWorld.productionsAll.push([a[0], new Decimal(a[1])]);
        else prod[1] = prod[1].plus(a[1]);
      });
      //  Starting
      w.startingUnit.forEach(a => {
        const un = retWorld.startingUnit.find(p => p[0].id === a[0].id);
        if (!un) retWorld.startingUnit.push([a[0], new Decimal(a[1])]);
        else un[1] = un[1].plus(a[1]);
      });
      //  Win
      w.winContidions.forEach(a => {
        const win = retWorld.winContidions.find(p => p.base.id === a.base.id);
        if (!win) retWorld.winContidions.push(new Price(a.base, a.price, 1));
        else win.price = win.price.plus(a.price.div(3));
      });
    });

    //  Unlocked
    retWorld.startingUnlocked = uniq(
      [].concat.apply([], worlds.map(w => w.startingUnlocked))
    );
    //  Malus
    retWorld.notWinConditions = uniq(
      [].concat.apply([], worlds.map(w => w.notWinConditions))
    );
    //  Bugs
    retWorld.additionalBugs = uniq(
      [].concat.apply([], worlds.map(w => w.additionalBugs))
    );

    return retWorld;
  }
  // tslint:disable-next-line:member-ordering
  static getRandomWorld(min: Decimal, max: Decimal, game: Game): World {
    const toMerge: World[] = [];
    if (
      game.allMateries.getSum(MasteryTypes.DOUBLE_PREFIX) > 0 &&
      Math.random() > 0.3
    ) {
      toMerge.push(sample(World.prefix));
    }
    toMerge.push(
      sample(World.prefix),
      sample(World.biome),
      sample(World.suffix)
    );
    if (
      game.allMateries.getSum(MasteryTypes.DOUBLE_SUFFIX) > 0 &&
      Math.random() > 0.3
    ) {
      toMerge.push(sample(World.suffix));
    }
    const world = World.merge(toMerge);
    world.setLevel(Utility.random(min, max), game);

    return world;
  }
}


src/app/model/units/malus-killer.ts
import { CONSTS } from "../CONSTATS";
import { FullUnit } from "../full-unit";
import { Game } from "../game";
import { Price } from "../price";
import { ProductionBonus } from "../production-bonus";
import { Research } from "../research";
import { UnitGroup } from "../unit-group";
import { World } from "../world";

export class MalusKiller extends UnitGroup {
  foodMalusKiller: FullUnit;
  woodMalusKiller: FullUnit;
  crystalMalusKiller: FullUnit;
  scienceMalusKiller: FullUnit;

  bonusRes: Research;

  private readonly price = new Decimal(500);

  constructor(game: Game) {
    super("Army", game);
  }
  declareStuff(): void {
    this.foodMalusKiller = new FullUnit("F");
    this.woodMalusKiller = new FullUnit("W");
    this.crystalMalusKiller = new FullUnit("C");
    this.scienceMalusKiller = new FullUnit("S");

    this.addUnits([
      this.foodMalusKiller,
      this.woodMalusKiller,
      this.crystalMalusKiller,
      this.scienceMalusKiller
    ]);

    this.bonusRes = new Research("aow", this.game.researches);
  }
  setRelations(): void {
    //  Production
    this.game.worldMalus.foodMalus1.addProducer(
      this.foodMalusKiller,
      new Decimal(-0.1)
    );
    this.game.worldMalus.woodMalus1.addProducer(
      this.woodMalusKiller,
      new Decimal(-0.1)
    );
    this.game.worldMalus.crystalMalus1.addProducer(
      this.crystalMalusKiller,
      new Decimal(-0.1)
    );
    this.game.worldMalus.scienceMalus1.addProducer(
      this.scienceMalusKiller,
      new Decimal(-0.1)
    );

    //  Buy actions
    this.foodMalusKiller.generateBuyAction([
      new Price(this.game.materials.food, this.price, 1.1)
    ]);
    this.woodMalusKiller.generateBuyAction([
      new Price(this.game.materials.food, this.price, 1.1)
    ]);
    this.crystalMalusKiller.generateBuyAction([
      new Price(this.game.materials.food, this.price, 1.1)
    ]);
    this.scienceMalusKiller.generateBuyAction([
      new Price(this.game.materials.food, this.price, 1.1)
    ]);

    const prodBon = new ProductionBonus(
      this.game.worldBonus.killBonus,
      new Decimal(0.1)
    );
    const prodBon2 = new ProductionBonus(this.bonusRes, new Decimal(0.1));
    this.bonusRes.prices = this.game.genSciencePrice(CONSTS.PRICE_2);

    //  Team and Twin
    this.list.forEach(u => {
      if (u instanceof FullUnit) {
        this.game.addTeamAction(u, CONSTS.TEAM_PRICE_2);
        this.game.addTwinAction(u, CONSTS.TWIN_PRICE_2);
        u.productionsAll.push(prodBon);
        u.productionsAll.push(prodBon2);
      }
    });
  }
  addWorlds(): void {
    const worldPre = new World("mkPre");
    const worldBio = new World("mkBio");
    const worldSuff = new World("mkSuff");

    [worldPre, worldBio, worldSuff].forEach(w => {
      w.productionsAll.push([this.game.worldBonus.killBonus, new Decimal(1)]);
    });

    worldPre.startingUnit.push([this.game.helpers.general, new Decimal(0)]);
    worldBio.startingUnit.push([this.game.helpers.headquarter, new Decimal(0)]);
    worldSuff.startingUnlocked.push(this.bonusRes);

    World.prefix.push(worldPre);
    World.biome.push(worldBio);
    World.suffix.push(worldSuff);
  }
}


src/app/model/strings.ts
export const STRINGS = {
  units: {
    //  Materials
    f: ["Food", "食物用于维持所有单位。"],
    w: ["Soil", "Soil is used to build better nest."],
    c: ["Crystal", "Crystal is used to make science."],
    s: ["Science", "Science is used to discover new things."],
    //  Ants
    l: ["Larva", "Larva"],
    q: ["Queen", "Queen"],
    n: ["Nest", "Nest"],
    //  Ghantering
    e: ["Drone", "Drone gather food."],
    h: ["Geologist", "Geologist will seek for crystals."],
    i: [
      "Student",
      [
        "Students will try to acquire science.",
        "Students were used to drink cristal."
      ]
    ],
    //  Workers
    a: [
      "Farmer",
      [
        "Hydroponics farmers uses crystals to grow fungi without soil.",
        "Ants are the World’s Best Fungus Farmers.",
        "Farming since 10.000 BC."
      ]
    ],
    b: ["Carpenter", "Carpenter ants builds nests in soil."],
    m: [
      "Miner",
      [
        "Miners extracts crystal through mining.",
        "Unsurprisingly, yet another idle game with miners."
      ]
    ],
    d: [
      "Scientist",
      [
        "Scientist gather science.",
        "Scientist were used to drink cristal.",
        "ScientAnts are like studentAnts, but betters."
      ]
    ],
    // Buildings
    aG: ["Farm", "An hydroponics farm."],
    bG: ["Soil Camp", "Soil Camp is used to get more soil."],
    mG: ["Mine", "A crystal mine."],
    dG: ["University", "University yeld scientists."],
    //  Engineers
    aGG: ["Hydro Eng.", "Hydro Engineers yeld Farm."],
    bGG: ["Soil Eng.", "Soil Engineers yeld Soil Camp."],
    mGG: ["Mine Eng.", "Mine Engineers yeld Mine."],
    dGG: ["Education Dep.", "Department of Education yeld University."],
    //  World Bonus
    1: ["Food Production Bonus", ""],
    2: ["Soil Production Bonus", ""],
    3: ["Crystal Production Bonus", ""],
    4: ["Science Production Bonus", ""],
    5: ["Kill Bonus", ""],
    larvaBon: ["Larvae Production Bonus", ""],
    queenBon: ["Queen Production Bonus", ""],
    //  Malus
    Mf1: ["Poisonous Fungus", "Poisonous Fungus"],
    Mf2: ["Ancient Fungus", "Ancient Fungus"],
    Mf3: ["Zombie Fungus", "Zombie Fungus"],
    Mw1: ["Carnivorous plant", ""],
    Mw2: ["Flypaper traps", ""],
    Mw3: ["Pitcher plant", ""],
    Mc1: ["Small Mites", ""],
    Mc2: ["Mites", ""],
    Mc3: ["Parasitic Mites", ""],
    Ms1: ["Zombie Fungus Spore", ""],
    Ms2: ["Zombie Fungus", ""],
    Ms3: ["Colony of Zombie Fungus ", ""],
    //  Killers
    F: ["Disinfestation Ant", "Destroy poisonous fungus."],
    W: ["Fire Ant", ["Burn Carnivorous plant", "Fire ant actually exists."]],
    C: ["Black garden ant", "Kill Mites."],
    S: ["Yellow crazy Ant", "Kill Zombie Fungus."],
    HB: ["General", "+10% killing."],
    HQ: ["Headquarter", "+10% killing."],
    //  Mastery Bonus
    scieMast: ["Science Mastery"],
    fooMast: ["Food Mastery"],
    wooMast: ["Wood Mastery"],
    cryMast: ["Crystall Mastery"],
    harvMast: ["Harvest Mastery"],
    matMast: ["Material Mastery"],
    armyMast: ["Army Mastery"],
    //  Supply
    fS: ["Food Supply", "Never ending food supply"],
    wS: ["Wood Supply", "Never ending wood supply"],
    cS: ["Crystal Supply", "Never ending crystall supply"],
    //  Bee
    Z: ["Foragging Bee", "Foragging Bee"],
    cb: ["Carpenter Bee", "Foragging Bee"],
    sb: ["Student Bee", "Foragging Bee"],
    Q: ["Bee Larva", "Bee Larva"],
    P: ["Bee Queen", ""],
    N: ["Bee Hive", ""],
    A: ["Bee Farmer", ""],
    B: ["Bee Carpenter", ""],
    E: ["Bee Miner", ""],
    D: ["Bee Scientist", ""],
    AG: ["2Bee Farmer", ""],
    BG: ["2Bee Carpenter", ""],
    EG: ["2Bee Miner", ""],
    DG: ["2Bee Scientist", ""],
    AGG: ["3Bee Farmer", ""],
    BGG: ["3Bee Carpenter", ""],
    EGG: ["3Bee Miner", ""],
    DGG: ["3Bee Scientist", ""],
    //  Wasp
    x: ["Foragging Wasp", "Foragging Wasp"],
    hw: ["Hornet", ["Hornet", "A wasp on steroids."]],
    sw: ["Smart Wasp", ""],
    y: ["Wasp Larva", ""],
    z: ["Wasp Queen", ""],
    r: ["Wasp Hive", ""],
    v: ["Wasp Farmer", ""],
    V: ["Wasp Carpenter", ""],
    o: ["Wasp Miner", ""],
    p: ["Wasp Scientist", ""],
    vG: ["2Wasp Farmer", ""],
    VG: ["2Wasp Carpenter", ""],
    oG: ["2Wasp Miner", ""],
    pG: ["2Wasp Scientist", ""],
    vGG: ["3Wasp Farmer", ""],
    VGG: ["3Wasp Carpenter", ""],
    oGG: ["3Wasp Miner", ""],
    pGG: ["3Wasp Scientist", ""],
    //  Helpers
    lf: ["Leaf Cutter", "Improve Farming."],
    hu: ["Hunter", ""],
    mw: ["Major Worker", ""],
    mh: ["Big Miner", "Improve Mining."],
    sh: ["Big Carpenter", "Improve soil gathering."]
  },
  researches: {
    t: ["Team Work", "Get production bonus based on manually bought."],
    T: ["Team Work 2", "Make team bonus upgradable."], // upgradable ?
    W: ["Twin", "Multiple units for same price."],
    r: ["Travel", "Travel to a new world !"],
    M: ["Mastery", "Get one mastery point."],
    h: ["Harvesting", "+100% resource from Gatherers."],
    1: ["1H Warp", "Free time warp."],
    2: ["2H Warp", "Free time warp."],
    3: ["3H Warp", "Free time warp."],
    SP: ["Spawn", "+50% larvae production."],
    //  Workers
    wo: ["Better Workers", "+30% production from workers."],
    we: ["Efficient Workers", "+10% resource gain."],
    w: ["Workers", "Workers"],
    a: [
      "Farmer",
      "Hydroponics farmers uses crystals to grow fungi without soil."
    ],
    b: [
      "Carpenter",
      "Carpenter ants build nests inside wood consisting of galleries."
    ],
    m: ["Miner", "miner"],
    d: ["Scientist", "scientist"],
    scie1: ["Scientific Method", "+50% science production"],
    // Buildings
    u: ["Buildings", "Buildings yields workers."],
    aG: ["Farm", "Hydroponics farm of Fungus."],
    bG: ["Soil Camp", "A fertile soil source."],
    mG: ["Mine", "A crystal mine."],
    dG: ["University", "University Yelds students."],
    //  Engineers
    e: ["Engineers", "Engineers yields Buildings."],
    aGG: ["Hydro Eng.", "Hydro Engineer yields Farm."],
    bGG: ["Soil Eng.", "Soil Engineer yields Soil Camp."],
    mGG: ["Mine Eng.", "Mine Engineer yields Mine."],
    dGG: ["Dep. of Education", "Department of Education yields University"],
    //  Bee
    A: ["Bee Farmer", "Farm flowers."],
    B: ["Bee Carpenter", "A Carpenter Bee."],
    E: ["Bee Miner", "A miner Bee."],
    D: ["Bee Scientist", "A scientist Bee."],
    AG: ["Flowers Farm", "A Farm of flowers."],
    BG: ["Bee Soil Camp", "A Bee Soil Camp"],
    EG: ["Bee Mine", "A Bee Mine"],
    DG: ["Bee University", "Bee University"],
    AGG: ["Florists Eng. ", "Florists Engineer yields Flowers Farms."],
    BGG: ["Bee Soil Eng.", "Bee Soil Engineer yields Bee Soil Camp."],
    EGG: ["Bee Mine Eng.", "Bee Mine Engineer yields Bee Mine."],
    DGG: ["Bee's Dep. of Education", "Bee's Department of Education"],
    // Wasp
    v: ["Wasp Farmer", "A Farming wasp."],
    V: ["Wasp Carpenter", "A Carpenter Wasp."],
    o: ["Wasp Miner", "A Miner Wasp."],
    p: ["Wasp Scientist", "A Scientist Wasp."],
    vG: ["Wasp Farm", "Waspy Farm."],
    VG: ["Wasp Soil Camp", "Waspy Soil Camp."],
    oG: ["Wasp Mine", "Waspy Mine."],
    pG: ["Wasp University", "Waspy University"],
    vGG: ["Wasp Hydro Eng.", "Wasp Hydro Engineer yields Wasp Farm."],
    VGG: ["Wasp Soil Eng.", "Wasp Soil Engineer yields Wasp Soil Camp Farm."],
    oGG: ["Wasp Mine Eng.", "Wasp Mine Engineer yields Wasp Mine."],
    pGG: [
      "Wasp's Dep. of Education",
      "Wasp's Department of Education yields Wasp University."
    ],
    //  Super Major
    sm: ["Super Major Helpers", ""],
    lf: ["Leaf Cutter", "Improve Farming."],
    mh: ["Big Miner", "Improve Mining."],
    sh: ["Big Carpenter", "Improve soil gathering."],
    //  Kill World
    aow: ["Art of War", "Improve killers by 10%"]
  },
  worlds: {
    "1Pre": "Hot",
    "2Pre": "Wooded",
    "3Pre": "Crystallized",
    "4Pre": "Scientific",
    "1Suf": "of Food",
    "2Suf": "of Wood",
    "3Suf": "of Crystall",
    "4Suf": "of Science",
    fBio: "Farm",
    wBio: "Forest",
    cBio: "Mine",
    sBio: "University",
    fSPre: "Food supply",
    wSPre: "Wood supply",
    cSPre: "Crystal supply",
    fSSuff: "with Food",
    wSSuff: "with Wood",
    cSSuff: "with Crystal",
    //  Bee
    beePre: "Apian",
    beeBio: "Hive",
    beeSuff: "of Bee",
    //  Wasp
    waspPre: "Waspy",
    waspBio: "Nest",
    waspSuff: "of Wasp",
    //  Super Major
    majorPre: "Giant",
    majorSuff: "of Super Major",
    //  Malus Bonus
    mkPre: "Hostile",
    mkBio: "Battlefield",
    mkSuff: "of War",
    //  Larva
    larvaPre: "Swarming",
    larvaSuff: "of Larvae",
    //  Queen
    queenPre: "Queenly",
    queenSuff: "of Queen"
  },
  prestige: {
    //  Followers 1
    e: ["Drone Follower", "Start new worlds with 5 more drones"],
    h: ["Geologist Follower", "Start new worlds with 5 more geologists"],
    i: ["Student Follower", "Start new worlds with 5 more students"],
    //  Followers 2
    a: ["Farmer Follower", "Start new worlds with 5 more farmer"],
    b: ["Carpenter Follower", "Start new worlds with 5 more farmer"],
    m: ["Miner Follower", "Start new worlds with 5 more miner"],
    d: ["Scientist Follower", "Start new worlds with 5 more scientist"],
    //  Time
    P: ["Time Producer", "+0.1 time/s"],
    B: ["Time Bank", "+1 max hour"],
    //  AutoBuy
    H: ["Hatch Auto Buyer", "Unlock Hatch Auto Buyer for all units"],
    t: ["Team Auto Buyer", "Unlock Team Auto Buyer for all units"],
    T: ["Twin Auto Buyer", "Unlock Twin Auto Buyer for all units"],
    M: ["Time Warp Auto Buyer", "Unlock one minute time warp Auto Buyer"],
    R: ["Research Auto Buyer", "Unlock Research Auto Buyer"],
    //  Tecnology
    F: ["Farming", "+10% food production"],
    C: ["Carpentery", "+10% wood production"],
    p: ["Mining", "+10% crystal production"],
    s: ["Studing", "+10% science production"],
    //  Team
    E: ["Team Work", "+30% Team Work Bonus"],
    //  Worlds
    w: ["Better World", "+10% World Bonus"],
    r: ["Better Army", "+10% Army killing stuff"],
    L: ["Max Level", "+10% Max World Level"]
  },
  actions: {
    warp60: ["Warp minute", "Time warp by minutes"],
    warp3600: ["Warp hour", "Time warp by hours"]
  },
  bug: {
    0: "Ants",
    1: "Bees",
    2: "Wasps",
    3: "Super Major"
  }
};
